### ISSUE TITLE
asyncssh drops connection after Public key authentication failure

### ISSUE THREAD

**Issue by silentnights** [ronf/asyncssh#233]
Hi Ron,

I am facing an authentication issue with the server. Basically I have a client that support ssh-agent authentication, which tries to authenticate via public key first and in case of failure it should try passowrd authentication. The client used Paramiko in the backend.

The problem I am currently having is that once the server receives `_USERAUTH_SERVICE` service, it set the `_next_service` to `None` and the authentication fails, the client sends a service request again with `_USERAUTH_SERVICE` but at this point the following check in `_process_service_request` fails:
`if service == self._next_service:`

And `DisconnectError` is raised which drops the connection with the client.

I think maybe since at that point `self._auth_in_progess` is `True` `send_userauth_failure()` should set `self._next_service` to `_USERAUTH_SERVICE` again?

All the above might also be nonsense and I am missing something. What do you think?

**Comment by ronf** [comment]
I don't believe it is correct for the client to ever send more than one service request for the _USERAUTH_SERVICE on a connection. That should only be sent once right after the key exchange completes. Once the server has sent back a MSG_SERVICE_ACCEPT, authentication begins, and that can include the client sending multiple MSG_USERAUTH_REQUEST messages, trying different forms of authentication until one of them succeeds (or the client gives up when it runs out of things to try).

Usually, the first MSG_USERAUTH_REQUEST sent by the client will be for an auth type of none, which is a way for the client to ask the sever what forms of auth it will accept. The server responds with a list, and the client can then pick one of the things in the server's list to try. On each auth failure, the server will update the list to reflect other methods the client can still try (which in some cases might still include the type it just tried).

**Comment by ronf** [comment]
I did a bit of digging through the SSH RFC (4253 in particular), and it looks like it doesn't really explicitly cover whether sending MSG_SERVICE_REQUEST multiple times is allowed or not. It's definitely not necessary, and from what I've seen this is never done by OpenSSH, but the only requirements I see in the RFC are:

  * After a key exchange with implicit server authentication, the client MUST wait for a response to its service request message before sending any further data.
  * MSG_SERVICE_REQUEST and MSG_SERVICE_ACCEPT MUST NOT be sent when a key exchange is in progress (between MSG_KEXINIT and MSG_NEWKEYS).

So, based on that, a server _could_ allow a client to send MSG_SERVICE_REQUEST for the ssh-userauth service multiple times, as long as the key exchange has completed. The client would need to wait for the MSG_SERVICE_ACCEPT before proceeding, which would introduce an unnecessary extra round trip into the process every time it did this, but there's nothing in the RFC that requires the additional MSG_SERVICE_REQUEST messages to be rejected.

Can you confirm that Paramiko actually sends MSG_SERVICE_REQUEST multiple times when trying out different authentication types, rather than just sending multiple MSG_USERAUTH_REQUEST messages to do this?

I wouldn't recommend setting self._next_service back to _USERAUTH_SERVICE only on auth failures. I think it would make more sense to just leave _next_service set to _USERAUTH_SERVICE on the server side until auth completes, removing setting it to None in _process_service_request and moving that down into send_userauth_success. In addition, I think there should be a check added in _process_service_request to only send the deferred packets when auth isn't already in progress. I haven't tested this, but I'm thinking of something like the following:

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index bc1971e..db58167 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -1298,6 +1298,7 @@ class SSHConnection(SSHPacketHandler):
         self._auth = None
         self._auth_in_progress = False
         self._auth_complete = True
+        self._next_service = None
         self.set_extra_info(username=self._username)
         self._send_deferred_packets()
 
@@ -1458,10 +1459,10 @@ class SSHConnection(SSHPacketHandler):
         if service == self._next_service:
             self.logger.debug2('Accepting request for service %s', service)
 
-            self._next_service = None
             self.send_packet(MSG_SERVICE_ACCEPT, String(service))
 
             if (self.is_server() and               # pragma: no branch
+                    not self._auth_in_progress and
                     service == _USERAUTH_SERVICE):
                 self._auth_in_progress = True
                 self._send_deferred_packets()
```

If this fixes the problem you are seeing with Paramiko, let me know, and I can make AsyncSSH more forgiving of these extra service request messages.

**Comment by silentnights** [comment]
Hi Ron,

I have been looking into the RFCs as well and you are right, there is no explicit mention if sending `_USERAUTH_SERVICE` is allowed or not, but yet I find the following paragraph: 

>    Message numbers used by services should be in the area reserved for
>    them (see [SSH-ARCH] and [SSH-NUMBERS]).  The transport level will
>    continue to process its own messages.

If I interpret it generally, it means all transport layer messages can be interpreted by the transport layer even after a service is started, that would include a request to new service like `_USERAUTH_SERVICE`

Again I am not expert and it is just my interpretation of the paragraph which might be wrong.

For your other question, I can confirm that paramiko sends several `_USERAUTH_SERVICE` after each authentication failure. I could even directly reproduce that by simply executing:

```python
client = SSHClient()
client.load_system_host_keys()
client.set_missing_host_key_policy(WarningPolicy())
client.connect(host,
               port=port,
               username=username,
               password=password,
               look_for_keys=False,
               timeout=120)
```

In case the ssh-agent is running and it provides keys for authentication, paramiko will attempt public key authentication and if that is not supported, it will send a new service_request for the next authentication which will be rejected now.

I tested you patch and it works. Thanks a lot for your awesome support. :-)

**Comment by ronf** [comment]
Happy to help - thanks for the confirmation about Paramiko's behavior!

I don't think it would make sense to allow a MSG_SERVICE_REQUEST asking for authentication after auth has completed. Part of the authentication request includes a service name itself ("ssh-connection"), and I don't think it would be a good idea for a server handling these connection requests to suddenly go back to accepting auth requests. It doesn't seem like Paramiko is doing that, though, and I see no problem allowing additional service requests for the _USERAUTH_SERVICE as long as authentication still hasn't completed.

I've gone ahead and checked the patch above into the "develop" branch in commit 1a2e14b, and this will be included in the next release.

**Comment by silentnights** [comment]
I agree with you. It does not make sense to authenticate twice on an ssh session. Do you have a rough idea when will the next release come out?

**Comment by ronf** [comment]
I've got enough built up to justify a new release at this point, so I'll try to have something out by this weekend.

**Comment by ronf** [comment]
These changes are now available in AsyncSSH 1.18.0.

**Comment by silentnights** [comment]
Thanks, updating soon.
